components/places/android/src/main/java/org/mozilla/places/Bookmarks.kt

```diff
@@ -0,0 +1,407 @@
+    override val type get() = BookmarkType.Folder
+}
+
+/**
+ * A bookmark which is a separator.
+ *
+ * It's type is always [BookmarkType.Separator], and it has no fields
+ * besides those defined by [BookmarkTreeNode].
+ */
+open class BookmarkSeparator(
+        override val guid: String,
+        override val dateAdded: Long,
+        override val lastModified: Long,
+        override val parentGUID: String?,
+        override val position: Int
+) : BookmarkTreeNode {
+    override val type get() = BookmarkType.Separator
+}
+
+/**
+ * The methods provided by a read-only or a read-write bookmarks connection.
+ */
+interface BookmarkReadInterface {
+    /**
+     * Returns the bookmark subtree rooted at `rootGUID`. This differs from
+     * `getBookmark` in that it populates folder children.
+     *
+     * Specifically, any [BookmarkFolder]s in the returned value will have their
+     * `children` list populated, and not just `childGUIDs` (Note: if
+     * `recursive = false` is passed, then this is only performed for direct
+     * children, and not for grandchildren).
+     *
+     * @param rootGUID the GUID where to start the tree. Defaults to `BookmarkRoot.Root`'s id,
+     *                 e.g. fetching the entire bookmarks tree.
+     *
+     * @param recursive Whether or not to return more than a single
+     *                  level of children for folders. If false, then
+     *                  any folders which are children of the requested
+     *                  node will *only* have their `childGUIDs`
+     *                  populated, and *not* their `children`. Defaults to
+     *                  true.
+     *
+     * @return The bookmarks tree starting at `rootGUID`, or null if the provided
+     *         id didn't refer to a known bookmark item.
+     */
+    fun getBookmarksTree(rootGUID: String = BookmarkRoot.Root.id, recursive: Boolean = true): BookmarkTreeNode?
+
+    /**
+     * Returns the information about the bookmark with the provided id. This differs from
+     * `getBookmarksTree` in that it does not populate the `children` list if `guid` refers
+     * to a folder (only it's `childGUIDs` list).
+     *
+     * @param guid the guid of the bookmark to fetch.
+     * @return The bookmark node, or null if the provided
+     *         guid didn't refer to a known bookmark item.
+     */
+    fun getBookmark(guid: String): BookmarkTreeNode?
+
+    /**
+     * Returns the list of bookmarks with the provided URL.
+     *
+     * Note that if the URL is not percent-encoded/punycoded, that will be performed
+     * internally, and so the returned bookmarks may not have an identical URL to
+     * the one passed in (however, it will be the same according to the
+     * [URL standard](https://url.spec.whatwg.org/)).
+     *
+     * @param url The url to search for.
+     * @return A list of bookmarks that have the requested URL.
+     */
+    fun getBookmarksWithUrl(url: String): List<BookmarkTreeNode>
```

Comment:
What's the intended use case? I'm assuming we'd mostly use this for looking up if a URL has been bookmarked already. A boolean result would be much more ergonomic in that case, and we can run a simpler query underneath to answer that question, as well. That is, I'm not sure why we'd need full information about all of the existing bookmarks for a given URL if we're just trying to determine if we should display a star or not in the UI. With the current API, we'd just see if the list isn't empty, and throw away the rest of the prepared data. That's a lot of wasted effort.

Reply:
This was requested by iOS.